### ISSUE TITLE
[CI] FrozenSearchableSnapshotsIntegTests testCreateAndRestorePartialSearchableSnapshot failing

### ISSUE THREAD

**Issue by bpintea** [elastic/elasticsearch#87501]
**Build scan:**
https://gradle-enterprise.elastic.co/s/nzkgnqet7vwt4/tests/:x-pack:plugin:searchable-snapshots:internalClusterTest/org.elasticsearch.xpack.searchablesnapshots.FrozenSearchableSnapshotsIntegTests/testCreateAndRestorePartialSearchableSnapshot

**Reproduction line:**
`./gradlew ':x-pack:plugin:searchable-snapshots:internalClusterTest' --tests "org.elasticsearch.xpack.searchablesnapshots.FrozenSearchableSnapshotsIntegTests.testCreateAndRestorePartialSearchableSnapshot" -Dtests.seed=AC36A259FD46B46A -Dtests.locale=no -Dtests.timezone=Hongkong -Druntime.java=18`

**Applicable branches:**
8.2

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.searchablesnapshots.FrozenSearchableSnapshotsIntegTests&tests.test=testCreateAndRestorePartialSearchableSnapshot

**Failure excerpt:**
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=44, name=elasticsearch[node_s0][masterService#updateTask][T#1], state=RUNNABLE, group=TGRP-FrozenSearchableSnapshotsIntegTests]

  at __randomizedtesting.SeedInfo.seed([AC36A259FD46B46A:D8B16388C0219A88]:0)

  Caused by: java.lang.AssertionError: (No message provided)

    at __randomizedtesting.SeedInfo.seed([AC36A259FD46B46A]:0)
    at org.elasticsearch.cluster.routing.allocation.IndexMetadataUpdater.applyChanges(IndexMetadataUpdater.java:134)
    at org.elasticsearch.cluster.routing.allocation.RoutingAllocation.updateMetadataWithRoutingChanges(RoutingAllocation.java:274)
    at org.elasticsearch.cluster.routing.allocation.AllocationService.buildResultAndLogHealthChange(AllocationService.java:150)
    at org.elasticsearch.cluster.routing.allocation.AllocationService.reroute(AllocationService.java:489)
    at org.elasticsearch.cluster.routing.BatchedRerouteService$1.execute(BatchedRerouteService.java:116)
    at org.elasticsearch.cluster.ClusterStateTaskExecutor$1.execute(ClusterStateTaskExecutor.java:95)
    at org.elasticsearch.cluster.service.MasterService.innerExecuteTasks(MasterService.java:901)
    at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:871)
    at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:253)
    at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:153)
    at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:114)
    at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:170)
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:714)
    at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:260)
    at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:223)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:833)

```

**Comment by bpintea** [comment]
Not same as #86186.

**Comment by original-brownbear** [comment]
This is caused by an overly optimistic assertion, fixed in 8.3+ already. I'll backport a fix to 8.2 as well.

**Comment by original-brownbear** [comment]
https://github.com/elastic/elasticsearch/pull/87649 fixes this
